bindings/matrix-sdk-ffi/src/client.rs

```diff
@@ -350,6 +353,68 @@ impl Client {
 }
 
 impl Client {
+    /// Requests the URL needed for login in a web view using OIDC. Once the web
+    /// view has succeeded, call `login_with_oidc_callback` with the callback it
+    /// returns.
+    pub(crate) async fn url_for_oidc_login(
+        &self,
+        oidc_configuration: &OidcConfiguration,
+    ) -> Result<Arc<OidcAuthorizationData>, AuthenticationError> {
+        let oidc_metadata: VerifiedClientMetadata = oidc_configuration.try_into()?;
+        let registrations_file = Path::new(&oidc_configuration.dynamic_registrations_file);
+        let static_registrations = oidc_configuration
+            .static_registrations
+            .iter()
+            .filter_map(|(issuer, client_id)| {
+                let Ok(issuer) = Url::parse(issuer) else {
+                    tracing::error!("Failed to parse {:?}", issuer);
+                    return None;
+                };
+                Some((issuer, ClientId(client_id.clone())))
+            })
+            .collect::<HashMap<_, _>>();
+        let registrations = OidcRegistrations::new(
+            registrations_file,
+            oidc_metadata.clone(),
+            static_registrations,
+        )?;
+
+        let data =
+            self.inner.oidc().url_for_oidc_login(oidc_metadata, registrations).await.map_err(
+                |e| match e {
+                    OidcError::MissingAuthenticationIssuer => AuthenticationError::OidcNotSupported,
+                    OidcError::MissingRedirectUri => AuthenticationError::OidcMetadataInvalid,
+                    _ => AuthenticationError::OidcError { message: e.to_string() },
+                },
+            )?;
+
+        Ok(Arc::new(data))
+    }
+
+    /// Completes the OIDC login process.
+    pub(crate) async fn login_with_oidc_callback(
+        &self,
+        authorization_data: Arc<OidcAuthorizationData>,
+        callback_url: String,
+    ) -> Result<(), AuthenticationError> {
+        let url = Url::parse(&callback_url).or(Err(AuthenticationError::OidcCallbackUrlInvalid))?;
+
+        self.inner.oidc().login_with_oidc_callback(&authorization_data, url).await.map_err(
+            |e| match e {
```

Comment:
If you `impl Into<AuthenticationError> for Error` then you could replace this whole `map_err` with just `?`, which might be easier to read.

Reply:
Oh yeah that would be nice. I've added a TODO for this as in the next PR I'll kill the `AuthenticationService` and simplify the error into an FFI `OidcError` so I'd rather not introduce breaking changes right now.